notary/src/tee.rs

```diff
@@ -185,46 +194,89 @@ pub async fn tee_proxy_service<S: AsyncWrite + AsyncRead + Send + Unpin>(
+}
+
+pub fn extract_request_and_response(
+  parsed_transcript: &[ParsedMessage],
+) -> Result<(ManifestRequest, ManifestResponse), NotaryServerError> {
+  let mut request_header = None;
+  let mut request_body = None;
+  let mut response_header = None;
+  let mut response_body = None;
+
+  // Classify parsed messages into headers and bodies
+  for ParsedMessage { payload, direction, .. } in parsed_transcript {
+    if let Some(app_data) = get_app_data(payload) {
+      debug!("App data message {:?}: {:?}", direction, String::from_utf8_lossy(&app_data));
+      match direction {
+        Direction::Sent if request_header.is_none() => request_header = Some(app_data),
+        Direction::Sent => request_body = Some(app_data),
+        Direction::Received if response_header.is_none() => response_header = Some(app_data),
+        Direction::Received => response_body = Some(app_data),
       }
     }
   }
 
-  if app_data_vec.len() != 3 {
-    return Err(NotaryServerError::MissingAppDataMessages(3, app_data_vec.len()));
-  }
+  // Ensure mandatory headers are present
+  let request_header =
+    request_header.ok_or(NotaryServerError::MissingAppDataMessages(Direction::Sent, 2, 0))?;
+  let response_header =
+    response_header.ok_or(NotaryServerError::MissingAppDataMessages(Direction::Received, 2, 0))?;
 
-  let request_header = app_data_vec[0].clone();
-  // TODO: Do we expect to get request_body as well part of app_data?
-  let response_header = app_data_vec[1].clone();
-  let response_body = app_data_vec[2].clone();
+  let request = ManifestRequest::from_payload(&request_header, request_body.as_deref())?;
+  let response =
+    ManifestResponse::from_payload(&response_header, response_body.as_deref().unwrap_or_default())?;
 
-  let request = ManifestRequest::from_payload(&request_header, None)?;
-  debug!("{:?}", request);
+  Ok((request, response))
+}
 
-  let response = ManifestResponse::from_payload(&response_header, &response_body)?;
-  debug!("{:?}", response);
+// TODO: Should TeeProof and other proofs be moved to `proofs` crate?
+// Otherwise, adding TeeProof::manifest necessitates extra dependencies on the client
+// Notice that all inputs to this function are from `proofs` crate
+pub fn create_tee_proof(
+  manifest: &Manifest,
+  request: &ManifestRequest,
+  response: &ManifestResponse,
+  State(state): State<Arc<SharedState>>,
+) -> Result<TeeProof, NotaryServerError> {
+  validate_notarization_legal(manifest, request, response)?;
+
+  let manifest_hash = manifest.to_keccak_digest()?;
+  let data = TeeProofData { manifest_hash: manifest_hash.to_vec() };
+
+  // TODO: That doesn't seem right
+  let to_sign = VerifyReply {
+    value:    format!("0x{}", hex::encode(manifest_hash)),
+    manifest: manifest.clone(),
+  };
+  let origo_sig = sign_verification(to_sign, State(state)).unwrap();
+  let as_bytes = hex::decode(&origo_sig.signature[2..]).unwrap();
+  let signature = hex::encode(&as_bytes);
+  Ok(TeeProof { data, signature })
```

Comment:
This is pretty close to what I was using for verification in the contracts for the origo flow. Is there a change we could match it up? Essentially we return this object here:

https://github.com/pluto/web-prover/blob/586c8abce67a15322ace9bd635f14d47577e37d9/client/src/origo.rs#L35

where the leaves are a 32byte hash of the value and a 32byte hash of the manifest, Then the digest is the root.

Reply:
Should this be the same struct, `SignedVerificationReply` for some reason (compatibility between modes, onchain verification, etc)?

Reply:
If we are able to keep it the same struct our on-chain logic will work with it which is nice. If we can use the same hashing and signing logic we used for the origo mode to populate the struct then it should work with the contract verification seamlessly

Reply:
So we should have the Tee proof return a `SignedVerificationReply` and the `Data`